warehouse/oidc/services.py

```diff
@@ -219,6 +220,23 @@ def _get_key_for_token(self, token):
         unverified_header = jwt.get_unverified_header(token)
         return self._get_key(unverified_header["kid"])
 
+    def token_identifier_exists(self, jti: str) -> bool:
+        """
+        Check if a JWT Token Identifier has already been used.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            return bool(r.exists(jti))
+
+    def store_jwt_identifier(self, jti: str, expiration: int) -> None:
+        """
+        Store the JTI with its expiration date if the key does not exist.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            # Defensive: to prevent races, we expire the JTI slightly after
+            # the token expiration date. Thus, the lock will not be
+            # released before the token invalidation.
+            r.set(jti, exat=expiration + 1, value="placeholder", nx=True)
```

Comment:
I've added a tiny offset here to make sure we don't encounter a situation where redis expires the key before the token invalidation (and thus, allowing to reuse it). WDYT?

(Just wanted to pinpoint it as the change is rather small and easy to miss)

Reply:
Makes sense to me. Local jitter can be more than 1s, though -- we could do 5s, since that's the standard-ish "leeway" that we use for TOTP expiries as well.

Reply:
Updated in 1d8156737366208cdf399